deeposlandia/datasets/tanzania.py

```diff
@@ -150,29 +193,121 @@ def _preprocess(self, image_filename, output_dir, labelling):
+            # randomly pick an image
+            x = np.random.randint(0, raw_img_width - self.image_size)
+            y = np.random.randint(0, raw_img_height - self.image_size)
+
+            tile_data = image_data[x:(x+self.image_size),
+                                   y:(y+self.image_size)]
+            tile_image = Image.fromarray(tile_data)
+            raster_features = get_image_features(raster)
+            tile_items = extract_tile_items(raster_features, labels,
+                                            x, y,
+                                            self.image_size,
+                                            self.image_size,
+                                            tile_srid=32737)
+            mask = self.load_mask(tile_items, raster_features, x, y)
+            label_dict = utils.build_labels(mask,
+                                            range(self.get_nb_labels()),
+                                            "tanzania")
+            labelled_image = utils.build_image_from_config(mask, self.labels)
+            if len(tile_items) > 0:
+                tiled_results = self._serialize(
+                    tile_image, labelled_image, label_dict,
+                    image_filename, output_dir, x, y, "nw"
+                )
+                result_dicts.append(tiled_results)
+                image_counter += 1
+                tile_image_ne = tile_image.transpose(Image.FLIP_LEFT_RIGHT)
+                labelled_image_ne = labelled_image.transpose(Image.FLIP_LEFT_RIGHT)
+                tiled_results_ne = self._serialize(
+                    tile_image_ne, labelled_image_ne, label_dict,
+                    image_filename, output_dir, x, y, "ne"
+                )
+                result_dicts.append(tiled_results_ne)
+                image_counter += 1
+                tile_image_sw = tile_image.transpose(Image.FLIP_TOP_BOTTOM)
+                labelled_image_sw = labelled_image.transpose(Image.FLIP_TOP_BOTTOM)
+                tiled_results_sw = self._serialize(
+                    tile_image_sw, labelled_image_sw, label_dict,
+                    image_filename, output_dir, x, y, "sw"
+                )
+                result_dicts.append(tiled_results_sw)
+                image_counter += 1
+                tile_image_se = tile_image_sw.transpose(Image.FLIP_LEFT_RIGHT)
+                labelled_image_se = labelled_image_sw.transpose(Image.FLIP_LEFT_RIGHT)
+                tiled_results_se = self._serialize(
+                    tile_image_se, labelled_image_se, label_dict,
+                    image_filename, output_dir, x, y, "se"
+                )
+                result_dicts.append(tiled_results_se)
+                image_counter += 1
+                del tile_image_se, tile_image_sw, tile_image_ne
+                del labelled_image_se, labelled_image_sw, labelled_image_ne
+            else:
+                if empty_image_counter < 0.1 * nb_images:
```

Comment:
maybe add a comment to clearly undertsand your intent

Reply:
I have completed the method docstring to explain the generation process in more details.